tests/base.py

```diff
@@ -0,0 +1,422 @@
+            for table in self.expected_metadata().keys()
+        }
+
+    def expected_replication_method(self):
+        """return a dictionary with key of table name nd value of replication method"""
+        return {table: properties.get(self.REPLICATION_METHOD, None)
+                for table, properties
+                in self.expected_metadata().items()}
+
+    def select_all_streams_and_fields(self, conn_id, catalogs, select_all_fields: bool = True):
+        """Select all streams and all fields within streams"""
+        for catalog in catalogs:
+            if catalog["tap_stream_id"] in self.expected_sync_streams():
+                schema = menagerie.get_annotated_schema(conn_id, catalog['stream_id'])
+
+                non_selected_properties = []
+                if not select_all_fields:
+                    # get a list of all properties so that none are selected
+                    non_selected_properties = schema.get('annotated-schema', {}).get(
+                        'properties', {}).keys()
+
+                connections.select_catalog_and_fields_via_metadata(
+                    conn_id, catalog, schema, [], non_selected_properties)
+
+    def run_and_verify_sync(self, conn_id):
+        """
+        Run a sync job and make sure it exited properly.
+        Return a dictionary with keys of streams synced and values of records synced for each stream
+        """
+        # Run a sync job using orchestrator
+        sync_job_name = runner.run_sync_mode(self, conn_id)
+
+        # Verify tap and target exit codes
+        exit_status = menagerie.get_exit_status(conn_id, sync_job_name)
+        menagerie.verify_sync_exit_status(self, exit_status, sync_job_name)
+
+        # Verify actual rows were synced
+        sync_record_count = runner.examine_target_output_file(
+            self, conn_id, self.expected_streams(), self.expected_primary_keys())
+        self.assertGreater(
+            sum(sync_record_count.values()), 0,
+            msg="failed to replicate any data: {}".format(sync_record_count)
+        )
+        print("total replicated row count: {}".format(sum(sync_record_count.values())))
+
+        return sync_record_count
+
+    def filter_output_file_for_records(self, output_file, stream_name):
+        return [message['data']
+                for message in output_file[stream_name]['messages']
+                if message['action'] == 'upsert']
+
+    def get_unique_records(self, stream, records):
+        unique_records = set()
+        for record in records:
+            # Build the primary key for this record, maintaining the same order for the fields
+            record_primary_key = [record[field]
+                                  for field in sorted(self.expected_primary_keys()[stream])]
+            # Cast to a tuple to make it hashable
+            unique_records.add(tuple(record_primary_key))
+        return unique_records
```

Comment:
I don't recognize these methods, are they necessary? I think the 2nd looks like it could alter records in a way that would invalidate assertions.

Reply:
Can we drop the filter output method, I don't think it is used now?